### ISSUE TITLE
CS_OUT arguments not assigned

### ISSUE THREAD

**Issue by jamesford42** [mono/CppSharp#438]
Trying to use CS_OUT on an argument that is a class marked with CS_API and passed in as a pointer (intending that the method call will dereference and assign a new value), results in the error:
Use of unassigned out parameter 'ExampleType'.

In particular this line is the problem. This testing of if the parameter is null prior to deciding to allocate a new one is invalid for an 'out' parameter in C#. 
```C#
var arg0 = packet == (ExampleType) null ? global::System.IntPtr.Zero : packet.__Instance;
```

**Comment by genuinelucifer** [comment]
I recreated the issue with :
https://gist.github.com/genuinelucifer/9e23701dee397e5c41c3
.
What i guess is both the functions should generate the same C# code for ref type class.. I am not clear as what a value type 'out' argument would act as! Also i am not completely sure about what should be in C++/CLI code for these.
Please guide me as to what should be the expected output in each case.

**Comment by tritao** [comment]
You're right, we should generate the same code for ref type class. Well it might have minor differences but it should mostly look the same.

And in C++/CLI case I think it needs to use a tracking reference (written as `%` type modifier). See 
https://msdn.microsoft.com/en-us/library/6f63s5b1.aspx.

**Comment by genuinelucifer** [comment]
Thanks. Will read about the tracking reference and then try to implement it.

**Comment by genuinelucifer** [comment]
I fixed the CSharp side of the code generation with :
https://gist.github.com/genuinelucifer/d73aa4039903b0c06e6d
.
I assume it was the expected solution.
.
But, i am NOT able to implement on the C++/CLI side because I am unable to figure out how the C++/CLI side is differentiating between "CS_OUT RefClass param" and "CS_OUT RefClass* param" when it is generating only "RefClass^% param" for both!
I just know that the difference comes at the point where param.Visit is called at line number 996 in CLISOurcesTemplate.cs
```C#
                var marshal = new CLIMarshalNativeToManagedPrinter(ctx);
                param.Visit(marshal);
```
Please provide furthur guidance on how to implement it,

**Comment by jamesford42** [comment]
Unforunately, I am not familiar C++/CLI so can offer no guidance there.

**Comment by tritao** [comment]
@genuinelucifer I didn't understand your question about CLI. What is the problem?

**Comment by genuinelucifer** [comment]
@tritao I went through the code to see where the first line of assignment is generated for one function and is not generated for other function. It was quite visible in CSharpTextTemplate.cs and hence the fix was easy.
However, in CLISourcesTemplate.cs, "param.Visit()" is called which makes all the difference... I tried to see the value of different values for both the functions and found that all the values passed in 'ctx' used to create 'marshal' are same for both the functions. Moreover, I couldn't actually figure out which function would get called on calling 'param.Visit()'... tried peek definition but it ends up to abstract function.
